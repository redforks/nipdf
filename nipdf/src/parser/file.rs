use super::{ws_terminated, FileError, ParseError, ParseResult};
use crate::{
    function::{Domain, Domains},
    object::{Dictionary, Entry, Frame, FrameSet, ObjectValueError, RuntimeObjectId, XRefSection},
    parser::{parse_dict, parse_indirect_stream},
};
use log::{error, info};
use memchr::memmem::rfind;
use nom::{
    branch::alt,
    bytes::complete::tag,
    character::complete::{char, satisfy, u32},
    combinator::{complete, map, map_res, opt, recognize},
    error::{context, ErrorKind, ParseError as NomParseError},
    multi::{count, fold_many1, many0},
    number::complete::{be_u16, be_u24, be_u32, be_u8},
    sequence::{preceded, separated_pair, tuple},
    InputIter, InputLength, InputTake, Parser, Slice,
};
use prescript::sname;
use std::{fmt::Display, ops::RangeFrom, str::from_utf8_unchecked};

/// Return `None`` if file not start with `%PDF-`
pub fn parse_header(buf: &[u8]) -> ParseResult<Option<&str>> {
    let one_digit = || satisfy(|c| c.is_ascii_digit());

    fn new_header(buf: &[u8]) -> std::result::Result<Option<&str>, FileError> {
        assert_eq!(3, buf.len());
        if buf[0] != b'1' {
            Err(FileError::UnsupportedVersion(
                String::from_utf8_lossy(buf).to_string(),
            ))
        } else {
            // safe to call from_utf8_unchecked(), because buf is checked for ascii digits
            Ok(Some(unsafe { from_utf8_unchecked(buf) }))
        }
    }

    let (buf, _) = tag("%")(buf)?;
    let (buf, v) = opt(tag("PDF-"))(buf)?;
    if v.is_none() {
        return Ok((buf, None));
    }

    map_res(
        recognize(tuple((one_digit(), char('.'), one_digit()))),
        new_header,
    )(buf)
}

/// Return start position of object tag from the end of the buffer.
/// Object tag occupies a whole line.
fn r_find_start_object_tag(mut buf: &[u8], tag: &[u8]) -> Option<usize> {
    fn is_new_line(buf: &[u8], pos: usize) -> bool {
        buf.get(pos).map_or(true, |ch| *ch == b'\n' || *ch == b'\r')
    }

    loop {
        match rfind(buf, tag) {
            None => return None,
            Some(pos) => {
                if is_new_line(buf, pos + tag.len()) && (pos == 0 || is_new_line(buf, pos - 1)) {
                    return Some(pos);
                } else {
                    buf = &buf[..pos];
                }
            }
        }
    }
}

fn new_r_to_tag<'a>(tag: &'static [u8]) -> impl FnMut(&'a [u8]) -> ParseResult<'a, ()> {
    move |buf| {
        let pos = r_find_start_object_tag(buf, tag);
        if let Some(pos) = pos {
            Ok((&buf[pos..], ()))
        } else {
            Err(nom::Err::Failure(ParseError::from_error_kind(
                buf,
                ErrorKind::Fail,
            )))
        }
    }
}

fn parse_trailer(buf: &[u8]) -> ParseResult<Dictionary> {
    preceded(ws_terminated(tag(b"trailer")), ws_terminated(parse_dict))(buf)
}

struct CrossReferenceStreamDict {
    size: u32,
    index: Option<Domains<u32>>,
    w: Vec<u32>,
}

impl CrossReferenceStreamDict {
    pub fn new(d: &Dictionary) -> Result<Self, ObjectValueError> {
        let size = d
            .get(&sname("Size"))
            .ok_or(ObjectValueError::DictKeyNotFound)?
            .int()? as u32;
        let index = d
            .get(&sname("Index"))
            .map(|o| Domains::<u32>::try_from(o).unwrap());
        let w = d
            .get(&sname("W"))
            .ok_or(ObjectValueError::DictKeyNotFound)?
            .arr()?
            .iter()
            .map(|o| o.int().map(|v| v as u32))
            .collect::<Result<Vec<_>, _>>()?;

        Ok(Self { size, index, w })
    }
}

/// Return nom parser to parse u32 value by byte length (0, 1, 2, 3, 4),
/// if n is 0, return parser takes 0 bytes and returns 0.
/// if n > 1, n32 stored in big endian n bytes.
fn segment_parser<'a, I: 'a>(n: u32) -> Box<dyn Parser<I, u32, nom::error::VerboseError<I>> + 'a>
where
    I: InputIter<Item = u8> + InputLength + InputTake + Slice<RangeFrom<usize>>,
{
    match n {
        0 => Box::new(|v| Ok((v, 0_u32))),
        1 => Box::new(map(be_u8, |v| v as u32)),
        2 => Box::new(map(be_u16, |v| v as u32)),
        3 => Box::new(be_u24),
        4 => Box::new(be_u32),
        _ => unreachable!(),
    }
}

/// Parse xref from cross-reference streams
fn parse_xref_stream(input: &[u8]) -> ParseResult<(XRefSection, Dictionary)> {
    fn to_parse_error<E: Display>(e: E) -> nom::Err<ParseError<'static>> {
        error!("should be xref table stream: {}", e);
        nom::Err::Error(ParseError::from_error_kind(b"", ErrorKind::Fail))
    }

    let (buf, stream) = parse_indirect_stream(input)?;
    let d = stream.as_dict();
    let d = CrossReferenceStreamDict::new(d).map_err(to_parse_error)?;

    let data = stream
        .decode_without_resolve_length(input, None)
        .map_err(to_parse_error)?;
    assert_eq!(3, d.w.len());
    let (a, b, c) = (d.w[0], d.w[1], d.w[2]);
    debug_assert_eq!(
        data.len() % (a + b + c) as usize,
        0,
        "stream data length should multiple of w0 + w1 + w2"
    );
    let (_, entries) = complete(count(
        move |i| tuple((segment_parser(a), segment_parser(b), segment_parser(c)))(i),
        data.len() / (a + b + c) as usize,
    ))(data.as_ref())
    .map_err(to_parse_error)?;

    let size = d.size;
    let sections = d
        .index
        .unwrap_or_else(|| Domains(vec![Domain::new(0, size)]));

    let mut r = Vec::with_capacity(sections.0.iter().map(|x| x.end as usize).sum());
    let mut entries = entries.into_iter();
    for Domain { start, end: size } in sections.0 {
        for (idx, (a, b, c)) in entries.by_ref().take(size as usize).enumerate() {
            let c: u16 = c.try_into().unwrap();
            let idx: u32 = idx.try_into().unwrap();
            match a {
                0 => r.push((start + idx, Entry::in_file(0, c, false))),
                1 => r.push((start + idx, Entry::in_file(b, c, true))),
                2 => r.push((start + idx, Entry::in_stream(RuntimeObjectId(b), c))),
                _ => info!("unknown xref stream entry type: {a}, idx: {idx}, ignored",),
            }
        }
    }

    Ok((buf, (r, stream.take_dict())))
}

// Assumes buf start from xref
fn parse_xref_table(buf: &[u8]) -> ParseResult<XRefSection> {
    let record_count_parser = context(
        "record count",
        ws_terminated(separated_pair(u32, tag(b" "), u32)),
    );
    let record_parser = context(
        "record",
        map(
            ws_terminated(tuple((
                u32,
                tag(b" "),
                u32,
                tag(b" "),
                alt((tag(b"n"), tag(b"f"))),
            ))),
            |(offset, _, generation, _, ty)| {
                // safe because clamp into u16 range
                #[allow(clippy::cast_possible_truncation)]
                Entry::in_file(
                    offset,
                    // saturation to u16 for incorrect pdf file, they may use 65536
                    // which out of the range of u16
                    generation.clamp(0, u16::MAX as u32) as u16,
                    ty == b"n",
                )
            },
        ),
    );
    let group = tuple((record_count_parser, many0(record_parser)));
    let parser = fold_many1(group, Vec::new, |mut table, ((start, count), entries)| {
        assert_eq!(count as usize, entries.len());
        for (i, entry) in entries.into_iter().enumerate() {
            table.push((start + u32::try_from(i).unwrap(), entry));
        }
        table
    });

    preceded(context("xref", ws_terminated(tag(b"xref"))), parser)(buf)
}

fn parse_startxref(buf: &[u8]) -> ParseResult<u32> {
    preceded(ws_terminated(tag(b"startxref")), ws_terminated(u32))(buf)
}

// Assumes buf start from xref
fn parse_frame(buf: &[u8]) -> ParseResult<(Dictionary, XRefSection)> {
    map(
        alt((
            tuple((
                context("xref table", parse_xref_table),
                context("trailer", parse_trailer),
            )),
            parse_xref_stream,
        )),
        |(xref_table, trailer)| (trailer, xref_table),
    )(buf)
}

pub fn parse_frame_set(input: &[u8]) -> ParseResult<FrameSet> {
    fn get_prev(frame: &Frame) -> Option<i32> {
        frame.trailer.get(&sname("Prev")).map(|o| o.int().unwrap())
    }

    let mut frames = Vec::new();
    let (buf, _) = context("move to xref", new_r_to_tag(b"startxref"))(input)?;
    let (_, pos) = context("locate frame pos", parse_startxref)(buf)?;
    info!("frame pos: {}", pos);
    let (_, frame) = parse_frame(&input[pos as usize..])?;
    let frame = Frame::new(pos, frame.0, frame.1);
    let mut prev = get_prev(&frame);
    frames.push(frame);

    while let Some(pos) = prev {
        info!("frame pos: {}", pos);
        let buf = &input[pos as usize..];
        let (_, frame) = parse_frame(buf)?;
        let frame = Frame::new(pos as u32, frame.0, frame.1);
        prev = get_prev(&frame);
        frames.push(frame);
    }

    Ok((&input[..0], frames))
}

#[cfg(test)]
mod tests;
