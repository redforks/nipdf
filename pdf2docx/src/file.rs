//! Contains types of PDF file structures.

use anyhow::Result as AnyResult;
use itertools::Itertools;
use nom::Finish;
use once_cell::unsync::OnceCell;
use pdf2docx_macro::pdf_object;
use std::collections::HashMap;

use crate::{
    object::{Dictionary, FrameSet, Name, Object, ObjectValueError, RootPdfObject, SchemaDict},
    parser::{parse_frame_set, parse_header, parse_indirected_object},
};
use log::error;

use nohash_hasher::BuildNoHashHasher;

mod page;
pub use page::*;

type IDOffsetMap = HashMap<u32, u32, BuildNoHashHasher<u32>>;

pub struct XRefTable<'a> {
    buf: &'a [u8],
    id_offset: IDOffsetMap, // object id -> offset
}

impl<'a> XRefTable<'a> {
    fn new(buf: &'a [u8], id_offset: IDOffsetMap) -> Self {
        Self { buf, id_offset }
    }

    #[cfg(test)]
    pub fn empty() -> Self {
        Self {
            buf: &[],
            id_offset: IDOffsetMap::default(),
        }
    }

    pub fn scan(frame_set: &FrameSet) -> IDOffsetMap {
        let mut r = IDOffsetMap::with_capacity_and_hasher(5000, BuildNoHashHasher::default());
        for (id, entry) in frame_set.iter().rev().flat_map(|f| f.xref_section.iter()) {
            if entry.is_used() {
                r.insert(*id, entry.offset());
            } else {
                r.remove(id);
            }
        }
        r
    }

    pub fn from_frame_set(buf: &'a [u8], frame_set: &FrameSet) -> Self {
        Self::new(buf, Self::scan(frame_set))
    }

    /// Return `buf` start from where `id` is
    pub fn resolve_object_buf(&self, id: u32) -> Option<&'a [u8]> {
        self.id_offset
            .get(&id)
            .map(|offset| &self.buf[*offset as usize..])
    }

    pub fn parse_object(&self, id: u32) -> Result<Object<'a>, ObjectValueError> {
        self.resolve_object_buf(id)
            .ok_or(ObjectValueError::ObjectIDNotFound)
            .and_then(|buf| {
                parse_indirected_object(buf)
                    .finish()
                    .map(|(_, o)| o.take())
                    .map_err(ObjectValueError::from)
            })
    }

    pub fn iter_ids(&self) -> impl Iterator<Item = u32> + '_ {
        self.id_offset.keys().copied()
    }

    pub fn count(&self) -> usize {
        self.id_offset.len()
    }
}

pub trait DataContainer<'a> {
    fn get_value(&self, key: &'a str) -> Option<&Object<'a>>;
}

impl<'a> DataContainer<'a> for Dictionary<'a> {
    fn get_value(&self, key: &'a str) -> Option<&Object<'a>> {
        debug_assert!(!key.starts_with('/'));
        self.get(&Name::borrowed(key.as_bytes()))
    }
}

/// Get value from first dictionary that contains `key`.
impl<'a> DataContainer<'a> for Vec<&Dictionary<'a>> {
    fn get_value(&self, key: &'a str) -> Option<&Object<'a>> {
        debug_assert!(!key.starts_with('/'));
        let key = Name::borrowed(key.as_bytes());
        for dict in self {
            if let Some(v) = dict.get(&key) {
                return Some(v);
            }
        }
        None
    }
}

pub struct ObjectResolver<'a> {
    xref_table: XRefTable<'a>,
    objects: HashMap<u32, OnceCell<Object<'a>>, BuildNoHashHasher<u32>>,
}

impl<'a> ObjectResolver<'a> {
    pub fn new(xref_table: XRefTable<'a>) -> Self {
        let mut objects =
            HashMap::with_capacity_and_hasher(xref_table.count(), BuildNoHashHasher::default());
        xref_table.iter_ids().for_each(|id| {
            objects.insert(id, OnceCell::new());
        });

        Self {
            xref_table,
            objects,
        }
    }

    #[cfg(test)]
    pub fn empty() -> Self {
        Self {
            xref_table: XRefTable::empty(),
            objects: HashMap::default(),
        }
    }

    #[cfg(test)]
    pub fn setup_object(&mut self, id: u32, v: Object<'a>) {
        self.objects.insert(id, OnceCell::with_value(v));
    }

    pub fn resolve_pdf_object<'b, T: RootPdfObject<'a, 'b>>(
        &'b self,
        id: u32,
    ) -> Result<T, ObjectValueError> {
        let obj = self.resolve(id)?.as_dict()?;
        T::new(id, obj, self)
    }

    pub fn opt_resolve_pdf_object<'b, T: RootPdfObject<'a, 'b>>(
        &'b self,
        id: u32,
    ) -> Result<Option<T>, ObjectValueError> {
        Self::to_opt(self.resolve_pdf_object(id))
    }

    /// Resolve object with id `id`, if object is reference, resolve it recursively.
    /// Return `None` if object is not found.
    pub fn opt_resolve(&self, id: u32) -> Result<Option<&Object<'a>>, ObjectValueError> {
        Self::to_opt(self.resolve(id))
    }

    /// Resolve object with id `id`, if object is reference, resolve it recursively.
    pub fn resolve(&self, id: u32) -> Result<&Object<'a>, ObjectValueError> {
        self.objects
            .get(&id)
            .ok_or(ObjectValueError::ObjectIDNotFound)?
            .get_or_try_init(|| {
                let mut o = self.xref_table.parse_object(id)?;
                loop {
                    match o {
                        Object::Reference(id) => {
                            let id = id.id().id();
                            o = self.xref_table.parse_object(id)?;
                        }
                        _ => break,
                    }
                }
                Ok(o)
            })
    }

    /// Resolve value from data container `c` with key `k`, if value is reference,
    /// resolve it recursively. Return `None` if object is not found.
    pub fn opt_resolve_container_value<'b: 'a, 'd: 'c, 'c, C: DataContainer<'a>>(
        &'d self,
        c: &'c C,
        id: &'b str,
    ) -> Result<Option<&'c Object<'a>>, ObjectValueError> {
        Self::to_opt(self.resolve_container_value(c, id))
    }

    /// Resolve value from data container `c` with key `k`, if value is reference,
    /// resolve it recursively.
    pub fn resolve_container_value<'b: 'a, 'd: 'c, 'c, C: DataContainer<'a>>(
        &'d self,
        c: &'c C,
        id: &'b str,
    ) -> Result<&'c Object<'a>, ObjectValueError> {
        self._resolve_container_value(c, id).map(|(_, o)| o)
    }

    pub fn opt_resolve_container_pdf_object<
        'b: 'a,
        'd: 'c,
        'c,
        C: DataContainer<'a>,
        T: RootPdfObject<'a, 'c>,
    >(
        &'d self,
        c: &'c C,
        id: &'b str,
    ) -> Result<Option<T>, ObjectValueError> {
        Self::to_opt(self.resolve_container_pdf_object(c, id))
    }

    pub fn resolve_container_pdf_object<
        'b: 'a,
        'd: 'c,
        'c,
        C: DataContainer<'a>,
        T: RootPdfObject<'a, 'c>,
    >(
        &'d self,
        c: &'c C,
        id: &'b str,
    ) -> Result<T, ObjectValueError> {
        let (id, obj) = self._resolve_container_value(c, id)?;
        let obj = obj.as_dict()?;
        T::new(id.expect("Should be root pdf object"), obj, self)
    }

    fn _resolve_container_value<'b: 'a, 'd: 'c, 'c, C: DataContainer<'a>>(
        &'d self,
        c: &'c C,
        id: &'b str,
    ) -> Result<(Option<u32>, &'c Object<'a>), ObjectValueError> {
        let obj = c.get_value(id).ok_or(ObjectValueError::ObjectIDNotFound)?;

        if let Object::Reference(id) = obj {
            self.resolve(id.id().id()).map(|o| (Some(id.id().id()), o))
        } else {
            Ok((None, obj))
        }
    }

    /// Resolve root pdf_objects from data container `c` with key `k`, if value is reference,
    /// resolve it recursively. Return empty vector if object is not found.
    /// The raw value should be an array of references.
    pub fn resolve_container_pdf_object_array<
        'b: 'a,
        'd: 'c,
        'c,
        C: DataContainer<'a>,
        T: RootPdfObject<'a, 'c>,
    >(
        &'d self,
        c: &'c C,
        id: &'b str,
    ) -> Result<Vec<T>, ObjectValueError> {
        let arr = c.get_value(id);
        arr.map_or_else(
            || Ok(vec![]),
            |arr| {
                let arr = arr.as_arr()?;
                let mut res = Vec::with_capacity(arr.len());
                for obj in arr {
                    let id = obj.as_ref()?;
                    let dict = self.resolve(id.id().id()).map(|o| o)?;
                    res.push(T::new(id.id().id(), dict.as_dict()?, self)?);
                }
                Ok(res)
            },
        )
    }

    fn to_opt<T>(o: Result<T, ObjectValueError>) -> Result<Option<T>, ObjectValueError> {
        o.map(Some).or_else(|e| {
            if let ObjectValueError::ObjectIDNotFound = e {
                Ok(None)
            } else {
                Err(e)
            }
        })
    }
}

#[pdf_object("Catalog")]
#[root_object]
trait CatalogDictTrait {
    #[typ("Name")]
    fn version(&self) -> Option<&str>;
    #[nested_root]
    fn pages(&self) -> PageDict<'a, 'b>;
}

#[derive(Debug)]
pub struct Catalog {
    id: u32,
    pages: Vec<Page>,
    ver: Option<String>,
}

impl Catalog {
    fn parse(id: u32, resolver: &mut ObjectResolver) -> Result<Self, ObjectValueError> {
        let catalog_dict: CatalogDict = resolver.resolve_pdf_object(id)?;
        let dict = resolver.resolve(id)?.as_dict()?;
        let dict = SchemaDict::new(dict, resolver, "Catalog")?;

        let root_page_id = dict.required_ref("Pages")?;
        let ver = catalog_dict.version().map(|s| s.to_owned());
        let pages = Page::parse(resolver.resolve_pdf_object(root_page_id)?, resolver)?;
        Ok(Self { id, pages, ver })
    }

    pub fn pages(&self) -> &[Page] {
        self.pages.as_slice()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Trailer<'a> {
    dict: Dictionary<'a>,
}

impl<'a> Trailer<'a> {
    pub fn new(dict: Dictionary<'a>) -> Self {
        Self { dict }
    }

    pub fn total_objects(&self) -> Result<i32, ObjectValueError> {
        self.dict
            .get(&Name::borrowed(b"Size"))
            .ok_or(ObjectValueError::DictNameMissing)
            .and_then(|obj| obj.as_int())
    }
}

#[derive(Debug)]
pub struct File {
    ver: String,
    total_objects: u32,
    catalog: Catalog,
}

#[derive(Debug, Clone, thiserror::Error)]
pub enum FileError {
    #[error("catalog object is required")]
    CatalogRequired,
    #[error("missing required trailer value")]
    MissingRequiredTrailerValue,
}

impl File {
    pub fn parse(buf: &[u8]) -> AnyResult<(Self, ObjectResolver)> {
        let (_, head_ver) = parse_header(buf).unwrap();
        let (_, frame_set) = parse_frame_set(buf).unwrap();
        let xref = XRefTable::from_frame_set(buf, &frame_set);
        let mut resolver = ObjectResolver::new(xref);

        let trailers = frame_set.iter().map(|f| &f.trailer).collect_vec();
        let root_id = trailers
            .iter()
            .find_map(|t| t.get(&Name::borrowed(b"Root")))
            .unwrap();
        let root_id = root_id.as_ref().unwrap().id().id();
        let catalog = resolver
            .resolve_container_value(&trailers, "Root")
            .map_err(|_| FileError::CatalogRequired)?;
        let ver = catalog
            .as_dict()?
            .get(&Name::borrowed(b"Version"))
            .map(|o| -> Result<String, ObjectValueError> { Ok(o.as_name().unwrap().to_owned()) })
            .unwrap_or(Ok(head_ver.to_owned()))?;
        let total_objects = resolver
            .resolve_container_value(&trailers, "Size")
            .map_err(|_| FileError::MissingRequiredTrailerValue)?
            .as_int()? as u32;
        let catalog = Catalog::parse(root_id, &mut resolver)?;

        Ok((
            Self {
                ver,
                total_objects,
                catalog,
            },
            resolver,
        ))
    }

    pub fn version(&self) -> &str {
        &self.ver
    }

    pub fn total_objects(&self) -> u32 {
        self.total_objects
    }

    pub fn catalog(&self) -> &Catalog {
        &self.catalog
    }
}

#[cfg(test)]
mod tests;
