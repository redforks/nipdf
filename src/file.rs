//! Contains types of PDF file structures.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct Header<'a>(&'a [u8]);

impl<'a> Header<'a> {
    pub fn new(buf: &'a [u8]) -> Self {
        Self(buf)
    }

    pub fn ver(&self) -> &str {
        std::str::from_utf8(self.0).unwrap()
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct Tail {
    xref_pos: u32,
}

impl Tail {
    pub fn new(xref_pos: u32) -> Self {
        Self { xref_pos }
    }

    pub fn xref_pos(&self) -> u32 {
        self.xref_pos
    }
}
