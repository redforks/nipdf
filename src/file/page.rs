use crate::object::{Array, Dictionary, ObjectValueError, SchemaDict};

use super::ObjectResolver;
use std::iter::once;

#[derive(Debug, Copy, Clone)]
pub struct Rectangle {
    pub left_x: f32,
    pub lower_y: f32,
    pub right_x: f32,
    pub upper_y: f32,
}

/// Convert from raw array, auto re-order to (left_x, lower_y, right_x, upper_y),
/// see PDF 32000-1:2008 7.9.5
impl<'a> From<&Array<'a>> for Rectangle {
    fn from(arr: &Array<'a>) -> Self {
        let mut iter = arr.iter();
        let left_x = iter.next().unwrap().as_number().unwrap();
        let lower_y = iter.next().unwrap().as_number().unwrap();
        let right_x = iter.next().unwrap().as_number().unwrap();
        let upper_y = iter.next().unwrap().as_number().unwrap();
        Self {
            left_x: left_x.min(right_x),
            lower_y: lower_y.min(upper_y),
            right_x: left_x.max(right_x),
            upper_y: lower_y.max(upper_y),
        }
    }
}

struct PageDict<'a, 'b> {
    d: SchemaDict<'a, 'b, [&'static str; 2]>,
}

impl<'a, 'b> PageDict<'a, 'b> {
    pub fn new(id: u32, dict: &'b Dictionary<'a>) -> Result<Self, ObjectValueError> {
        let d = SchemaDict::new(id, dict, ["Pages", "Page"])?;
        Ok(Self { d })
    }

    pub fn is_leaf(&self) -> bool {
        self.d.type_name() == "Page"
    }

    pub fn parent_id(&self) -> Option<u32> {
        self.d.opt_int("Parent").unwrap().map(|id| id as u32)
    }

    pub fn kids(&self) -> Vec<u32> {
        self.d
            .opt_arr_map("Kids", |o| Ok(o.as_ref()?.id().id()))
            .unwrap()
            .unwrap_or_default()
    }

    pub fn media_box(&self) -> Option<Rectangle> {
        self.d.opt_rectangle("MediaBox").unwrap()
    }

    pub fn crop_box(&self) -> Option<Rectangle> {
        self.d.opt_rectangle("CropBox").unwrap()
    }
}

#[derive(Debug)]
pub struct Page {
    /// pdf object id
    id: u32,
    media_box: Rectangle,
    crop_box: Option<Rectangle>,
}

impl Page {
    pub fn id(&self) -> u32 {
        self.id
    }

    pub fn media_box(&self) -> Rectangle {
        self.media_box
    }

    pub fn crop_box(&self) -> Option<Rectangle> {
        self.crop_box
    }

    /// Parse page tree to get all pages
    pub fn parse<'a, 'b>(
        root_id: u32,
        resolver: &'b ObjectResolver<'a>,
    ) -> Result<Vec<Page>, ObjectValueError> {
        let mut pages = Vec::new();
        let mut parents = Vec::new();
        fn handle<'a, 'b, 'c>(
            id: u32,
            resolver: &'b ObjectResolver<'a>,
            pages: &'c mut Vec<Page>,
            parents: &'c mut Vec<PageDict<'a, 'b>>,
        ) -> Result<(), ObjectValueError> {
            let d = resolver.resolve(id).unwrap();
            let d = PageDict::new(id, d.as_dict()?)?;
            if d.is_leaf() {
                pages.push(Page::from_leaf(id, &d, &parents[..]));
            } else {
                let kids = d.kids();
                parents.push(d);
                for kid in kids {
                    handle(kid, resolver, pages, parents)?;
                }
            }
            Ok(())
        }
        handle(root_id, resolver, &mut pages, &mut parents)?;
        Ok(pages)
    }

    fn from_leaf<'a, 'b>(id: u32, d: &PageDict<'a, 'b>, parents: &[PageDict<'a, 'b>]) -> Self {
        let media_box = once(d)
            .chain(parents.iter())
            .map(|d| d.media_box())
            .find_map(|r| r)
            .unwrap();
        let crop_box = once(d)
            .chain(parents.iter())
            .map(|d| d.crop_box())
            .find_map(|r| r);

        Self {
            id,
            media_box,
            crop_box,
        }
    }
}

#[cfg(test)]
mod tests;
