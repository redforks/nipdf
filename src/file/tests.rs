use crate::object::Object;

use std::path::PathBuf;

use test_case::test_case;

use super::*;

#[test_case(None => None)]
#[test_case(Object::Name(Name::borrowed(b"abc")) => Some("abc".into()))]
fn catalog_ver(ver: impl Into<Option<Object<'static>>>) -> Option<String> {
    let ver = ver.into();
    let mut dict = Dictionary::default();
    dict.insert("Type".into(), "/Catalog".into());
    if let Some(ver) = ver {
        dict.insert(Name::borrowed(b"Version"), ver);
    }
    let cat = CatalogDict::new(dict);
    cat.ver().map(|v| v.into())
}

#[test]
fn xref_table_resolve_object_buf() {
    let buf = b"1234567890";
    let mut id_offset = IDOffsetMap::default();
    id_offset.insert(1, 5);
    id_offset.insert(2, 3);
    let xref_table = XRefTable::new(buf, id_offset);

    assert_eq!(xref_table.resolve_object_buf(1), Some(&b"67890"[..]));
    assert_eq!(xref_table.resolve_object_buf(2), Some(&b"4567890"[..]));
    assert_eq!(xref_table.resolve_object_buf(3), None);
}

#[test]
fn object_resolver() {
    let buf = b"   2 0 obj 5 endobj 1 0 obj null endobj 3 0 obj 2 0 R endobj";
    let mut id_offset = IDOffsetMap::default();
    id_offset.insert(1, 20);
    id_offset.insert(2, 3);
    id_offset.insert(3, 40);
    let xref_table = XRefTable::new(buf, id_offset);
    let resolver = ObjectResolver::new(xref_table);

    assert_eq!(resolver.resolve(1), Ok(&Object::Null));
    assert_eq!(resolver.resolve(2), Ok(&Object::Integer(5)));
    assert_eq!(resolver.resolve(3), Ok(&Object::Integer(5)));
    assert_eq!(resolver.resolve(1), Ok(&Object::Null));
}

#[test]
fn parse_file() {
    let mut p = PathBuf::from(file!());
    assert_eq!(
        p.pop().then(|| p.pop().then(|| p.pop())).flatten(),
        Some(true)
    );
    p.push("sample_files");
    p.push("normal");
    p.push("SamplePdf1_12mb_6pages.pdf");
    let buf = std::fs::read(p).unwrap();
    let (f, _) = File::parse(&buf).unwrap();
    assert_eq!("1.5", f.version());
    assert_eq!(311, f.total_objects);
}
