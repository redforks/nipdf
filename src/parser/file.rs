use nom::{
    bytes::complete::tag,
    character::complete::u32,
    character::complete::{char, line_ending, satisfy},
    combinator::{complete, map, map_res, recognize},
    sequence::{preceded, terminated, tuple},
};

use crate::file::{Header, Tail};

use super::{FileError, ParseError, ParseResult};

fn parse_header(buf: &[u8]) -> ParseResult<'_, Header<'_>> {
    let one_digit = || satisfy(|c| c.is_digit(10));

    fn new_header(buf: &[u8]) -> std::result::Result<Header<'_>, FileError> {
        assert_eq!(3, buf.len());
        if buf[0] != b'1' {
            Err(FileError::UnsupportedVersion(
                String::from_utf8_lossy(buf).to_string(),
            ))
        } else {
            Ok(Header::new(buf))
        }
    }

    terminated(
        preceded(
            tag("%PDF-"),
            map_res(
                recognize(tuple((one_digit(), char('.'), one_digit()))),
                new_header,
            ),
        ),
        line_ending,
    )(buf)
}

/// Reverse iterator of lines, ignore empty lines.
fn riter_lines(buf: &[u8]) -> impl Iterator<Item = &[u8]> + '_ {
    // must ignore empty lines, because the way we detect the end of a line is
    // check either '\r' or '\n', for '\r\n' or '\n\r' we will get extra empty line.
    let mut iter = buf.rsplit(|&c| c == b'\n' || c == b'\r');
    iter.filter(|&line| !line.is_empty())
}

fn parse_tail(buf: &[u8]) -> ParseResult<Tail> {
    let mut iter = riter_lines(buf);
    let line = iter.next().unwrap_or_default();
    complete(tag("%%EOF"))(line)?;
    let line = iter.next().unwrap_or_default();
    map(complete(u32), Tail::new)(line)
}

#[cfg(test)]
mod tests;
