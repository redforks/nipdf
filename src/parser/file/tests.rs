use super::*;
use crate::file::Header;
use insta::assert_debug_snapshot;
use test_case::test_case;

#[test]
fn parse_file_header() {
    let buf = b"%PDF-1.7\n%comment\n";
    let (buf, header) = parse_header(buf).unwrap();
    assert_eq!(buf, b"%comment\n");
    assert_eq!(header, Header::new(b"1.7"));

    let buf = b"%PDF-1.7";
    let err = parse_header(buf);
    assert!(err.is_err());
}

#[test_case(None, b"hello", b"world"; "not exist")]
#[test_case(Some(0), b"hello", b"hello"; "matches")]
#[test_case(Some(1), b"\nhello", b"hello"; "after newline")]
#[test_case(Some(1), b"\nhello\n", b"hello"; "end with newline")]
#[test_case(Some(2), b"\r\nhello\r\n", b"hello"; "CRLF")]
#[test_case(Some(4), b"foo\nfoo\nbar", b"foo"; "from end")]
#[test_case(None, b"abc-foo", b"foo"; "not the whole line")]
fn test_after_tag_r(exp: Option<usize>, buf: &[u8], tag: &[u8]) {
    assert_eq!(exp, r_find_start_object_tag(buf, tag));
}

#[test]
fn test_parse_tail() {
    let buf = b"\nstartxref\n1234\n%%EOF";
    assert_eq!((b"".as_slice(), Tail::new(1234)), parse_tail(buf).unwrap());
}

#[test]
fn test_parse_trailer() {
    let buf = b"trailer\n<< /Size 1 >>\nstartxref\n1234\n%%EOF";
    assert_debug_snapshot!(parse_trailer(buf).unwrap());
}

#[test_case(b"xref\n1 0\n"; "empty")]
#[test_case(b"xref\n1 2\n0000000000 00000 n \n0000000010 00000 n \n"; "two entries")]
fn test_parse_xref_table(buf: impl AsRef<[u8]>) {
    assert_debug_snapshot!(parse_xref_table_section(buf.as_ref()).unwrap());
}
