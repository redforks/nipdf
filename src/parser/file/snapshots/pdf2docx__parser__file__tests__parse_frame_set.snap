---
source: src/parser/file/tests.rs
expression: "parse_frame_set(b\"%PDF-1.7\nxref\n1 1\n0000000000 00000 n\ntrailer\n<< /Size 1 >>\nstartxref\n9\n%%EOF\nxref\n1 1\n0000000000 00000 n\ntrailer\n<< /Prev 9 >>\nstartxref\n77\n%%EOF\n\").unwrap()"
---
(
    [],
    FrameSet(
        [
            Frame {
                xref_pos: 77,
                trailer: Trailer {
                    dict: {
                        Name(
                            [
                                47,
                                80,
                                114,
                                101,
                                118,
                            ],
                        ): Integer(
                            9,
                        ),
                    },
                },
                xref_table: XRefTable {
                    entries: {
                        1: XRefEntry(
                            0,
                            0,
                            true,
                        ),
                    },
                },
            },
            Frame {
                xref_pos: 9,
                trailer: Trailer {
                    dict: {
                        Name(
                            [
                                47,
                                83,
                                105,
                                122,
                                101,
                            ],
                        ): Integer(
                            1,
                        ),
                    },
                },
                xref_table: XRefTable {
                    entries: {
                        1: XRefEntry(
                            0,
                            0,
                            true,
                        ),
                    },
                },
            },
        ],
    ),
)
